### DISCUSSION TITLE
Appcast parser rewritten to Swift

### DISCUSSION THREAD

**Discussion by jozefizso** [sparkle-project/Sparkle#2459] [Ideas]
Hi,
we are using Sparkle to update our application which integrated into Microsoft PowerPoint. Our standalone application can update using the `Sparkle.framework` but we need also update notifications inside PowerPoint.

As our library must be a `.dylib` to run inside PowerPoint, we cannot reference the full `Sparkle.framework` in our dynamic library target.

As we need to parse the `appcast.xml` file and get information if a new update is available, we need the `Appcast Support` part of the Sparkl project.

Would you be willing to accept PR which rewrites Appcast code to Swift?

Here is a sample code: https://github.com/slidoapp/Sparkle/tree/dev/appcast_swift_versions
And with new unit tests: https://github.com/slidoapp/Sparkle/tree/dev/appcast_swift_versions_tests

**Comment by zorgiepoo** [comment]
> we are using Sparkle to update our application which integrated into Microsoft PowerPoint. Our standalone application can update using the Sparkle.framework but we need also update notifications inside PowerPoint.

> As our library must be a .dylib to run inside PowerPoint, we cannot reference the full Sparkle.framework in our dynamic library target.

How does PowerPoint's or Office's plugin architecture work? If you can't bundle multiple resources it would then seem to not be very good. Is it already out-of-process from the main app or is it in-process? Sparkle's stance is to [discourage injecting Sparkle into host applications](https://sparkle-project.org/documentation/bundles/) that provide plug-in support and perhaps use an external tool process instead (like sparkle-cli) if needed. If multiple plug-ins use Sparkle (with the same class prefixes) injected in the same process, it's problematic.

Generally I don't want the Sparkle framework to use Swift (this is completely unrelated with how you're using Sparkle above) at least not until the deployment target is 10.14.6 or above (otherwise the framework will be bloated to include swift libraries). I don't think rewriting parts of Sparkle in Swift for the heck of it is a good justification for bumping the deployment target. Swift Package Manager sometimes causes issues when mixing Obj-C/Swift code so that would need to be assessed as well (we do provide a binary package though, so maybe the situation is different). I care about small code generation too so that would also need to be tested. Obj-C is also currently a more stable binary target (see Swift issues in https://github.com/apple/swift/issues/68163 for example where sometimes binary compatibility breaks).

**Reply by zorgiepoo** [reply]
Also you can dlopen a dynamic framework (`Foo.framework/Foo`) just like you can a dylib (`foo.dylib`) so I'm not sure it's actually true you're restricted to a vanilla dylib.

**Reply by jozefizso** [reply]
Yes, they use `dlopen` to load the addin extensions. The addin runs in-process. For simplicity of deployment we chose to use just a `.dylib` and not a framework.

We need only the update check to run in the PowerPoint, the whole update is implemented in our standalone app. As we parse the `appcast.xml` in side PowerPoint and determine there is a new version, we will call our app from command line to do the update.

**Reply by zorgiepoo** [reply]
I see. If you can call out to your app anyway, another approach is adding a CLI mode to your app that uses Sparkle to check and return if a new update is available (with [checkForUpdateInformation](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdateInformation)) similar to `sparkle-cli --probe`

Taking parts of the appcast/downloader code for your other process is not recommended or officially supported from Sparkle's point of view.

**Reply by jozefizso** [reply]
I am exploring the use of `sparkle-cli` but it is not available when we reference Sparkle using Swift PM. It is not included in `Sparkle-for-Swift-Package-Manager.zip` neither.

**Reply by zorgiepoo** [reply]
I would just create a CLI mode in your app. It needs to create an updater (you could use `SPUStandardUpdaterController` still), call [checkForUpdateInformation](https://sparkle-project.org/documentation/api-reference/Classes/SPUUpdater.html#/c:objc(cs)SPUUpdater(im)checkForUpdateInformation) immediately after starting the updater, exit by the time [-updater:didFinishUpdateCycleForUpdateCheck:error:](https://sparkle-project.org/documentation/api-reference/Protocols/SPUUpdaterDelegate.html#/c:objc(pl)SPUUpdaterDelegate(im)updater:didFinishUpdateCycleForUpdateCheck:error:) is called, and set up a runloop. The user driver used won't matter if you exit() by then and do the update check immediately after starting the updater.

Refer to [sparkle-cli code](https://github.com/sparkle-project/Sparkle/blob/2.x/sparkle-cli/main.m#L192) as an example of this. sparkle-cli itself is a very light wrapper around using Sparkle's framework.
